README.txt
				$Id: README,v 1.3 2000/01/03 08:56:55 js Exp $


UQWK
                       Copyright     1993-1994,             steve belczyk
                       Modifications 1996, 1997, 1998, 1999,
                                     2000                   jeroen scheerder


uqwk is a program which collects all a user's unread mail or news
and formats it into a packet for offline reading.  QWK, Simple
Offline Usenet Packet (SOUP), and ZipNews packet formats are supported.
See the manual page for sommand line options, environment variables and
examples.

Uqwk also accepts reply packets, so replies can be mailed or posted,
depending whether the message is marked private (email) or public (news).

Uqwk also supports a small offline command language, so the contents
of the user's .newsrc file can be viewed and manipulated offline.

0.  Install uqwk if not installed already; see the INSTALL file.

1.  Now you can try it.  Login as a normal user who has some mail.
    First, tell uqwk to print its options: issue

	uqwk -p

    Check that the output reflects your choices in the makefile.
    Next, issue:

	uqwk +r +m

    (The "+r" stops uqwk from clearing the user's mail spool file.)
    This should create a number of *.MSG files in your home directory,
    along with a file named AREAS.  These are the files the offline
    SOUP reader will need.  (Some readers expect these files to have been
    archived using an archiver like zip, lharc, or arj.  You may need
    to obtain Unix versions of these archivers.)

2a. If you will not be using uqwk in QWK mode (hurray!), you need not
    even  bother reading [2b]; advance to [3].
2b. If you are using uqwk to create QWK packets (please *don't*, really,
    since the QWK format does not allow storing Usenet messages
    correctly): QWK packets contain the name, location, etc., of the BBS
    from which messages are being downloaded.  If you plan to use the
    QWK format (please, please, please don't!), you should configure
    this information. The best way is probably to use environment
    variables.  If you are using a Bourne shell, you should add
    something like this to /etc/profile or .profile:

        UQ_BBS_NAME="My Super BBS"
	UQ_BBS_CITY="Somewhere, PA"
	UQ_BBS_PHONE="+1 215 555 1212"
	UQ_BBS_SYSOP="Joe Shmoe"
	UQ_BBS_ID="0,MYBBS"
	export UQ_BBS_NAME UQ_BBS_CITY UQ_BBS_PHONE
	export UQ_BBS_SYSOP UQ_BBS_ID

    If you use a C type shell, try something like this in your
    .cshrc or .login:

        setenv UQ_BBS_NAME "My Super BBS"
	setenv UQ_BBS_CITY "Somewhere, PA"
	setenv UQ_BBS_PHONE "+1 215 555 1212"
	setenv UQ_BBS_SYSOP "Joe Shmoe"
	setenv UQ_BBS_ID "0,MYBBS"

    In both cases, the last entry, the "BBS ID", is the most important.
    It always consists of an integer, a comma, and a string less than
    nine characters, with no intervening spaces.  The string will be
    used to identify reply packets.

3.  Now would be a good time for reading the man page ("man uqwk")
    and the Frequently Asked Questions list ("more FAQ").


HISTORY.txt
$Id: HISTORY,v 1.13 2001/10/25 20:22:19 js Exp $

uqwk 2.21 (js, 10/25/2001)

- Rewrote a little part of the signal handling/error returning code

uqwk 2.20 (js, 10/24/2001)

- More bugfixed, most related to maildir and signal handling,
  as proposed by Jos en Bekker.

uqwk 2.19 (js, 10/23/2001)

- Fixed a bug when SUBSCRIBEing to a group, an offline command;
  with no connection to the server, this would still result
  in attempts to communicate with it.  Thanks, Jos den Bekker!

uqwk 2.18 (js, 10/05/2001)

- Fixed a bug where uqwk closed an already closed file handle,
  causing SIGSEGV's on some platforms, thanks to Konstantin
  Isakov and Peter Karlsson.

uqwk 2.17 (js, 04/19/2001)

- Fixed a one-off error in a mail reading error message, thanks to
  Jos den Bekker.

uqwk 2.16 (js, 06/21/2000)

- Fixed a bug that broke compilation without NNTP support, as noted
  by Peter Karlsson.  Thanks, Peter!

uqwk 2.15 (js, 01/03/2000)

- Applied Doug Kaufman's patches.  Thanks, Doug!

uqwk 2.14 (js, 10/19/1999)

- Included Helmut Karlowski's modifications to ignore errors on opening
  articles.  Thanks, Helmut!
- Brought the manual page up to date.

uqwk 2.13 (js, 09/11/1999)

- Fixed an NNTP command that accidentally carried a newline too much
  (causing existing groups to be reported as nonexistent in the worst
  case).  Thanks to Jeff Schroeder.

uqwk 2.12 (js, 07/15/1999)

- Fixed a nasty bug in SLNP (SOUP) mail reply handling in which article
  length was determined incorrectly

uqwk 2.11 (js, 05/20/1999)

- Several small changes to make uqwk compiler cleaner than ever
  (thanks mainly to George Schwarz)

uqwk 2.10 (js, 05/14/1999)

- Put under `configure' control
- Made to compile cleanly (even when being pedantic)

uqwk 2.09 (js, 05/12/1998)

- More code cleanup
- Preempted nonexisting MSG files from addition to the AREAS file (SOUP mode)
  (thanks to George Schwarz for pointing this out and testing)

uqwk 2.08 (js, 03/07/1998)

- Finished(?) maildir support; thanks to Jos den Bekker for code
  revision, contribution and testing.

uqwk 2.07 (js, 03/04/1998)

- Rudimentary maildir support
- Preparations for full maildir support
- Some cleanup

uqwk 2.06 (js, 06/02/1998)

- Removed obsolete (unused) code from misc.c

uqwk 2.05 (js, 05/25/1998)

- Tiny modification (due to a change in GNKSAlib's API)

uqwk 2.04 (js, 03/26/1998)

- fixes a summary mode bug (news.c) that somehow crept in along
  the way (thanks to Doug Kaufman for the report and the fix).

uqwk 2.03 (js, 02/16/1998)

- fixed an (ancient, yet unnoticed all this time...) bug in appending
  messages (both mail and news) to QWK "messages.dat" files.

uqwk 2.02 (js, 02/09/1998)

- fixed a tiny bug in init.c/close.c: when processing replies only,
  NNTP connections were opened/closed unnecessarily (NNTP uqwk only)

uqwk 2.01 (js, 01/06/1998)

- added basic status reporting when fetching mail

uqwk 2.0 (js, 12/16/1997 -- final version)

- updated to reflect MANKSA name (and interface) change
  (MANKSA is now (again) GNKSA, the Good Net-Keeping Seal of Approval)
- extensive acceptance testing in real life situations, no problems
  found -- so it went final.  Finally.

uqwk 2.0fc4 (js, 11/12/1997 -- fourth final candidate version)

- rnews batches (== SOUP newsgroup files) contain `indexes'
  which are _almost_ the article byte length -- almost, because
  line termination counts as a single byte, even if represented
  by multiple.
  When news articles -- they shouldn't, but sometimes do somehow --
  contain \r\n line termination, uqwk used a wrong index (the
  file size of the article instead of the size with the number
  of \r's subtracted).
  uqwk now makes excruciatingly sure that articles are written to
  the SOUP files in Unix format, with '\n' line termination,
  and correct article indexing in them.
- a few other (minor) fixes in news.c's error handling

uqwk 2.0fc2 (js, 10/22/1997 -- second final candidate version)

- changed (enhanced) control over message validation (-Vn switch)
- worked around a few bugs occurring in a particular Unix version
  better left unmentioned.

uqwk 2.0fc1 (js, 10/21/1997 -- first final candidate version)

- A few (very minor) cleanups in reply.c
- Rethought some of the GNKSA validation decisions
  (and adapted the man page accordingly)

uqwk 2.0b7 (js, 10/13/1997 -- seventh beta version)

- Updated calls to gnksalib routines

uqwk 2.0b6 (js, 10/07/1997 -- sixth beta version)

- Fixed (usually spurious) extra `group' (directly after FindActive()
  in DoGroup in NNTP mode
- Fixed one more instance of a conference being created before testing
  whether the group actually exists
- Reorganized/cleaned up a bit in news.c

uqwk 2.0b5 (js, 10/06/1997 -- fifth beta version)

- SOUP mode: a .MSG file was created, even for non-existing groups -- fixed
- fixed a fclose() on a possibly null filehandle (crashing bug)
- Updated the man page, and the getsoup script, for decent mailbox locking
  strategies, and some information about it

uqwk 2.0b4 (js, 10/06/1997 -- fourth beta version)

- Small robustness patch, and enhanced user feedback in the
  offline command handling functions (once again, thanks Jos den Bekker)

uqwk 2.0b3 (js, 10/03/1997 -- third beta version)

- More fixes in the mail handling part (again, thanks to Jos den Bekker)

uqwk 2.0b2 (js, 10/01/1997 -- second beta version)

- Fixes in offline.c's OffLine function (reported and fixed
  by Jos den Bekker <josdb@xs4all.nl>

uqwk 2.0b1 (js, 09/30/1997 -- first beta version)

- Calls sendmail directly when mailing, in _all_ cases (sometimes,
  "mail" was used, sometimes sendmail)
- Fixed a few bugs w.r.t. returning errorcodes
- Added `-I/path/to/inews' switch and global variables containing
  the inews path as well as a full inews call (/path/to/inews -h);
  removed `INEWS_WRAPPER' compile option.  Added UQ_INEWS_PATH
  environment variable
- Added (basic) message validation support (GNKSA, using gnksalib)

source/uqwk.h
/*
 *  $Id: uqwk.h,v 1.9 1999/10/19 08:48:42 js Exp $
 */

#ifdef	HAVE_CONFIG
  #include HAVE_CONFIG
#else
  #include "config-generic.h"
#endif

#include "uqwk-version.h"

#ifdef NNTP_USE_ACTIVE
	#undef	NNTP
	#define	NNTP
	#define NNTP_ACTIVESTR	"+full-active"
#else
	#define	NNTP_ACTIVESTR	""
#endif

#ifdef NNTPAUTH
	#undef	NNTP
	#define	NNTP
	#define	NNTPAUTHSTR	"auth"
#else
	#define	NNTPAUTHSTR	""
#endif

#ifdef NNTP
	#define NNTPSTR	"+nntp"
#else
	#define NNTPSTR	""
#endif

#ifdef SCORES
	#define SCORESTR	"+scoring"
#else
	#define SCORESTR	""
#endif

#ifdef GNKSA
	#define	NO_SIZE_T	/* don't ask... */
	#define GNKSASTR	"+gnksa"
	#define	GNKSA_VERBOSE	1
	#define GNKSA_HONOUR	2
#else
	#define GNKSASTR	""
#endif

#define	UQWK_VERSION	VERSION"/"CONFIGNAME""NNTPSTR""NNTPAUTHSTR""GNKSASTR""SCORESTR


#ifdef ALLOCATE
#define EXTERN
#else
#define EXTERN extern
#endif

#define	PATH_LEN	(256)	/* Length for file names, etc. */
#define BUF_LEN		(16384)	/* Length of general purpose buffer */

#define QWK_EOL		(227)	/* QWK end-of-line */
#define QWK_ACT_FLAG	(225)	/* Active message flag */
#define QWK_PUBLIC	' '	/* Public message flags */
#define QWK_PUBLIC2	'-'
#define QWK_PRIVATE	'+'	/* Private message flags */
#define QWK_PRIVATE2	'*'

#define SLNP_MAILER	SENDMAIL" -t"
#define ZIP_MAILER	SENDMAIL
#define XPRT_MAILER	SENDMAIL" -t"

#define MAIL_CONF_NAME  "Email"

#ifdef SCORES
  #define GLOBAL_SCORE_FILE "global"
#endif

/* Defaults */
#define	DEF_DO_DEBUG	(0)
#ifdef NNTPAUTH
  #define DEF_DO_AUTH	(0)
  #define DEF_AUTH_USER	""
  #define DEF_AUTH_PASS	""
#endif
#define DEF_DO_MAIL	(0)
#define DEF_DO_NEWS	(0)
#ifdef SCORES
  #define DEF_DO_KILLS	(1)
#endif
#define DEF_IGNORE0FD	(1)
#define DEF_INC_HDRS	(1)
#define DEF_PRT_OPTS	(0)
#define DEF_READ_ONLY	(0)
#define DEF_WAF_MODE	(0)
#define DEF_SLNP_MODE	(1)
#define DEF_ZIP_MODE	(0)
#define DEF_XRF_MODE	(1)
#define DEF_BW_KLUDGE	(0)
#define DEF_XPRT_MODE	(0)
#define DEF_SUM_MODE	(0)
#define DEF_SEL_MODE	(0)
#define DEF_EVERY_MODE	(0)
#define DEF_MAX_BLKS	(4096)	/* Exactly half a megabyte */
#define DEF_GRP_LEN	(15)
#define DEF_HOME_DIR	"."
#define DEF_USER_NAME	"nobody"
#define DEF_MAIL_FILE	"unknown"
#define DEF_NRC_FILE	"unknown"
#define DEF_REP_FILE	"none"
#define DEF_TRN_FILE	"none"
#define DEF_HOST_NAME	"nowhere"
#define DEF_SUM_FILE	"none"
#define DEF_SEL_FILE	"none"
#define DEF_NG_FILE	"none"
#ifdef SCORES
  #define DEF_KILL_DIR	".uqwk"
#endif

/* Runtime options */
EXTERN	int do_debug;
#ifdef NNTPAUTH
  EXTERN	int do_auth;		/* Do NNTP auth? */
#endif
EXTERN	int do_mail;			/* Process mail? */
EXTERN	int do_news;			/* Process news? */
#ifdef SCORES
  EXTERN	int do_kills;		/* Process kill/score files? */
  EXTERN	int kill_thresh;	/* global kill threshold */
  EXTERN	int group_kill_thresh;	/* kill threshold within this group */
#endif
EXTERN int ignore_0fd;

EXTERN	int inc_hdrs;			/* Include headers in messages? */
EXTERN	int prt_opts;			/* Display options; no processing */
EXTERN	int read_only;			/* Don't rewrite mail and .newsrc */
EXTERN	int max_blks;			/* Maximum blocks per QWK packet */
EXTERN	int grp_len;			/* Maximum newsgroup name length */
EXTERN	int waf_mode;			/* Waffle type .newsrc */
EXTERN	int slnp_mode;			/* Write SLNP packet */
EXTERN	int zip_mode;			/* Write ZipNews packet */
EXTERN	int xrf_mode;			/* Cancel crosspostings */
EXTERN	int bw_kludge;			/* Blue Wave kludge */
EXTERN	int xprt_mode;			/* Expert mode */
EXTERN	int sum_mode;			/* Summary mode */
EXTERN	int sel_mode;			/* Selection mode */
EXTERN	int every_mode;			/* Write every grp to CONTROL.DAT */

#ifdef NNTPAUTH
	EXTERN	char auth_user[PATH_LEN];/* NNTP auth user */
	EXTERN	char auth_pass[PATH_LEN];/* NNTP auth pass */
#else
	EXTERN	char act_file[PATH_LEN];	/* Active file */
	EXTERN	char news_dir[PATH_LEN];	/* News spool dir */
#endif
EXTERN	char mail_file[PATH_LEN];	/* mail spool */
EXTERN	char mail_dir[PATH_LEN];	/* dir for mail spool */
EXTERN	char home_dir[PATH_LEN];	/* home directory */
EXTERN	char user_name[PATH_LEN];	/* user's login name */
EXTERN	char bbs_name[PATH_LEN];	/* BBS name */
EXTERN	char bbs_city[PATH_LEN];	/* BBS city */
EXTERN	char bbs_phone[PATH_LEN];	/* BBS phone number */
EXTERN	char bbs_sysop[PATH_LEN];	/* BBS sysop name */
EXTERN	char bbs_id[PATH_LEN];		/* BBS ID */
EXTERN	char nrc_file[PATH_LEN];	/* .newsrc file */
EXTERN	char rep_file[PATH_LEN];	/* Reply packet file name */
EXTERN	char trn_file[PATH_LEN];	/* Newsgroup name translation table */
EXTERN	char host_name[PATH_LEN];	/* Host machine name */
EXTERN	char sum_file[PATH_LEN];	/* Summary file */
EXTERN	char sel_file[PATH_LEN];	/* Selection file */
EXTERN	char ng_file[PATH_LEN];		/* Desired ng file */

EXTERN	char inews_path[PATH_LEN];		/* path to inews */
EXTERN	char posting_agent[PATH_LEN];		/* posting agent (inews + '-h') */

#ifdef SCORES
	EXTERN  char kill_dir[PATH_LEN];/* directory for score/kill files */
#endif

#ifdef	GNKSA
EXTERN	int gnksa_level;			/* Validate messages before posting? */
#endif

char *getenv();
char *Fgets();
char *ParseFrom();
struct act_ent *FindActive();
struct sub_ent *SubList();
struct sub_ent *MarkRead();
struct sub_ent *FixSub();
struct conf_ent *NewConference();

/* Various globals */
EXTERN	char *progname;			/* Program name */
EXTERN	int msg_cnt;			/* Total number of messages */
EXTERN	int conf_cnt;			/* Total number of conferences */
EXTERN	FILE *msg_fd;			/* MESSAGES.DAT file desc */
EXTERN	char msg_fn[PATH_LEN];		/* SLNP *.MSG file name */
EXTERN	FILE *ctl_fd;			/* CONTROL.DAT file desc */
EXTERN	FILE *ndx_fd;			/* xxx.NDX file desc */
EXTERN	char ndx_fn[PATH_LEN];		/* xxx.NDX file name */
EXTERN	FILE *act_fd;			/* Active file file desc */
EXTERN	FILE *nrc_fd;			/* .newsrc file desc */
EXTERN	FILE *rep_fd;			/* Reply packet file desc */
EXTERN	FILE *nws_fd;			/* ZipNews .nws file desc */
EXTERN	FILE *idx_fd;			/* ZipNews .idx file desc */
EXTERN	FILE *mai_fd;			/* ZipNews .mai file desc */
EXTERN	int zip_flag;			/* idx entry written for this ng? */
EXTERN  char buf[BUF_LEN];	        /* General purpose buffer */
EXTERN	int blk_cnt;			/* Blocks written to messages.dat */
EXTERN	int sum_flag;			/* Summary hdr for this ng yet? */
EXTERN	FILE *sum_fd;			/* Summary file desc */
EXTERN	FILE *sel_fd;			/* Selection file desc */

/* This is the stuff we remember about each spooled mail message */
EXTERN	struct mail_ent
{
	long begin;		/* Offset of start of header */
	long text;		/* Offset to end of header, start of text */
	long end;		/* Offset to start of next message */
	struct mail_ent *next;	/* Pointer to next */
} *mail_list;

/* This is stuff we remember about each "conference" */
EXTERN	struct conf_ent
{
	char *name;		/* Conference name */
	int number;		/* Conference number */
	int count;		/* Number of articles in this conference */
	struct conf_ent *next;	/* Pointer to next */
} *conf_list, *last_conf;

/* This is the QWK message header format */
struct qwk_hdr
{
        char status;
        char number[7];
        char date[8];
        char time[5];
        char to[25];
        char from[25];
        char subject[25];
        char password[12];
        char refer[8];
        char blocks[6];
        char flag;
        char conference[2];
        char msg_num[2];
        char tag;
};

EXTERN	struct qwk_hdr rep_hdr;		/* Header for replies */

/* Stuff we remember about each active newsgroup */
EXTERN	struct act_ent
{
	char *name;		/* Newsgroup name */
	int hi;			/* High article number */
	int lo;			/* Low article number */
	struct act_ent *next;	/* Pointer to next */
} *act_list;

/* Stuff we remember about the .newsrc file */
EXTERN	struct nrc_ent
{
	char *name;		/* Newsgroup name */
	int subscribed;		/* Subscribed flag */
	struct sub_ent *sub;	/* Subscription list */
	int conf;		/* Corresponding conference number */
	struct nrc_ent *next;	/* Pointer to next */
} *nrc_list;

/* Stuff we remember about the newsgroup name translation table */
EXTERN	struct trn_ent
{
	char *old;		/* Old name */
	char *new;		/* New name */
	struct trn_ent *next;	/* Pointer to next */
} *trn_list;

/* Stuff we remember about a subscription list (in the .newsrc) */
struct sub_ent
{
	int lo, hi;		/* Range of articles */
	struct sub_ent *next;	/* Pointer to next */
};

/* Stuff we remember about the desired newsgroup list */
EXTERN	struct ng_ent
{
	char *name;		/* Newsgroup name */
	struct ng_ent *next;	/* Pointer to next */
} *ng_list;

/* Function prototypes */

#include	<stdio.h>
#include	<stdlib.h>
#include	<string.h>
#include	<sys/types.h>
#include	<unistd.h>

void DefaultOptions();
void EnvOptions();
void CommandOptions(int, char **);
void InitStuff();
int DoReply();
int DoNews();
int DoMail();
int DoClassicMail();
int DoMaildirMail();
void AddMessage(FILE *fd, struct mail_ent *me, char *xhdrs);
void SLNPMessage(FILE *fd, struct mail_ent *me, char *xhdrs);
void ZipMessage(FILE *fd, struct mail_ent *me, char *xhdrs);
void QwkMessage(FILE *fd, struct mail_ent *me, char *xhdrs);
void DoFromLine(long off);
int CloseStuff();
int BadFlag (char *);
void PrintOptions();
int ReadTrans();
int ReadNG();
void OutOfMemory();
void MakeMailList();
void OpenZipMail();
void inttoms(int, unsigned char[4]);
int  buftoint(unsigned char c[4]);
void Spaces(char *, int);
void PadNum(int, char *, int);
void IntNum(int, char[2]);
void PadString( char *, char *, int);
void ParseDate(char *, struct qwk_hdr *);
char *ParseFrom(char *);
int WriteControl();
int WriteNewsrc();
int WriteAreas();
void TransName (char *);
void WriteSub (FILE *fd, struct sub_ent *sp);
int OpenZipFiles();
int ReadActive();
int ReadNewsrc();
int DoSelect();
int WriteJoin();
int DoGroup (struct nrc_ent *);
void NdxClose (FILE *);
void MsgClose (FILE *);
int IsRead (int, struct sub_ent *);
int DoArticle (char *, int, struct nrc_ent *, struct conf_ent *);
int DoXref (FILE *, long);
int SLNPArticle (FILE *, int);
int ZipArticle (FILE *, int, struct nrc_ent *);
int SumArticle (FILE *, int, long, struct nrc_ent *);
void ws(FILE *fd, struct sub_ent *, struct sub_ent *);
int Xref(char *);
int GetHdr (FILE *, char *, int, char *);
int DoZipReplies();
int SLNPReply();
int SendReply();
int SendWarning(char *);
int CheckTo(char *, int);
int SLuMail(char *);
int SLuNews(char *);
int SLbMail(char *);
int SLbNews(char *);
int SLPipe(FILE *, int, char *);
void AppendBadArticle(char *);
int ZipId();
int ZipMail(char *);
int ZipNews(char *);


int OffLine(FILE *, int);
int QWKOffLine (int, FILE *);
int GetLine(int *, FILE *);
void Help(FILE *);
int Subscribe(FILE *);
int Unsubscribe(FILE *);
int Groups(FILE *);
int Allgroups(FILE *);
int Shell(FILE *);
int Catchup(FILE *);

FILE *getactive_nntp();
void put_server ();
